### PR TITLE
new tutorials, sampledatasets

### PR DESCRIPTION
**Pull request by ycq091044**

### CODE REVIEW

File: pyhealth/datasets/base_dataset.py [hunk 1]
```diff
@@ -593,3 +603,170 @@ def task_stat(self) -> str:
...
+            a list of sample indices. Default is None.
+        visit_to_index: Optional[Dict[str, List[int]]], a dict mapping visit_id to a
+            list of sample indices. Default is None.
+    """
+
+    def __init__(self, samples, dataset_name="dataset"):
+        self.dataset_name: str = dataset_name
+        self.samples: Optional[List[Dict]] = samples
+        self.patient_to_index: Optional[Dict[str, List[int]]] = None
+        self.visit_to_index: Optional[Dict[str, List[int]]] = None
```
Comment (zzachw) [inline pyhealth/datasets/base_dataset.py:637]: I think you should call _index_patient() and _index_visit() in the init. You can also add some sanity check set_task() in the BaseDataset.

File: pyhealth/datasets/base_dataset.py [hunk 2]
```diff
@@ -593,3 +603,170 @@ def task_stat(self) -> str:
...
+    Args:
+        samples: the processed data samples.
+            E.g., 
+                samples[0] = {
+                    visit_id: 1,
+                    patient_id: 1,
+                    "codnition": ["A", "B"],
+                    "procedure": ["C", "D"],
+                    "label": 1
+                }
```
Comment (zzachw) [inline pyhealth/datasets/base_dataset.py:621]: Can you elaborate on the format of the samples? I think you can move the description in Attributes/samples here.

File: pyhealth/datasets/base_dataset.py [hunk 3]
```diff
@@ -593,3 +603,170 @@ def task_stat(self) -> str:
...
+                of sample indices.
+        """
+        visit_to_index = {}
+        for idx, sample in enumerate(self.samples):
+            visit_to_index.setdefault(sample["visit_id"], []).append(idx)
+        return visit_to_index
+
+
+    # TODO: check this
+    def get_all_tokens(self, key: str, sort: bool = True) -> List[str]:
```
Comment (zzachw) [inline pyhealth/datasets/base_dataset.py]: Can you update this function to match BaseDataset.get_all_tokens()?

File: pyhealth/datasets/base_dataset.py (hunk 1 above)
Comment (zzachw) [inline pyhealth/datasets/base_dataset.py]: I think you should call _index_patient() and _index_visit() in the init. You can also add some sanity check set_task() in the BaseDataset.

File: pyhealth/datasets/base_dataset.py [hunk 4]
```diff
@@ -593,3 +603,170 @@ def task_stat(self) -> str:
...
+                    patient_id: 1,
+                    "codnition": ["A", "B"],
+                    "procedure": ["C", "D"],
+                    "label": 1
+                }
+
+    Attributes:
+        samples: Optional[List[Dict]], a list of samples, each sample is a dict with
+            patient_id, visit_id, and other task-specific attributes as key.
+            Default is None.
```
Comment (zzachw) [inline pyhealth/datasets/base_dataset.py]: Can you elaborate on the format of the samples in Args? I think you can move the description in Attributes/samples here.

File: pyhealth/datasets/base_dataset.py [hunk 5]
```diff
@@ -593,3 +603,170 @@ def task_stat(self) -> str:
...
+    def __len__(self):
+        """Returns the number of samples in the dataset."""
+        return len(self.samples)
+
+    def stat(self) -> None:
+        """Prints some statistics of the dataset."""
+        self.task_stat()
+
+    # TODO: check this
+    def task_stat(self) -> None:
```
Comment (zzachw) [inline pyhealth/datasets/base_dataset.py]: Can you update this function to match BaseDataset.task_stat()? I think the get_label_distribution can be removed.

**Review by zzachw — CHANGES_REQUESTED** [review]

File: pyhealth/trainer.py [hunk 6]
```diff
@@ -272,6 +270,23 @@ def evaluate(self, dataloader) -> Dict[str, float]:
...
+        Args:
+            dataloader: Dataloader for evaluation.
+
+        Returns:
+            scores: a dictionary of scores.
+        """
+        y_true_all, y_prob_all, loss_mean = self.inference(dataloader)
+        y_true_all = np.concatenate(y_true_all, axis=0)
+        y_prob_all = np.concatenate(y_prob_all, axis=0)
+
```
Comment (zzachw) [inline pyhealth/trainer.py:288]: These two lines seem unnecessary

File: pyhealth/trainer.py [hunk 7]
```diff
@@ -245,17 +245,15 @@ def train(
...
-    def evaluate(self, dataloader) -> Dict[str, float]:
-        """Evaluates the model.
+    def inference(self, dataloader) -> Dict[str, float]:
+        """ Model inference.
 
         Args:
             dataloader: Dataloader for evaluation.
 
         Returns:
             scores: a dictionary of scores.
```
Comment (zzachw) [inline pyhealth/trainer.py]: Returns need update

**Review by zzachw — CHANGES_REQUESTED** [review]

**Review by zzachw — APPROVED** [review]

**Comment by jalengg** [comment]
Fixed in commit 2d4fbbdb. MANIFEST.in added to include all YAML config files in package distribution. Verified that all 8 config files are now included in the built package.

**Pull request merged**